### ISSUE TITLE
/api/v1/label/.../values get incorrect result when ...=__name__

### ISSUE THREAD

**Issue by n4mine** [VictoriaMetrics/VictoriaMetrics#343]
**Describe the bug**
/api/v1/label/.../values get nothing when ...=`__name__`, see following `step 3`

**To Reproduce**
1. get all series
```
$ curl -s  "127.0.0.1:8481/select/0/prometheus/api/v1/series" -d 'match[]={__name__=~".*"}' | jq .
{
  "status": "success",
  "data": [
    {
      "__name__": "m_v",
      "t1": "v1"
    }
  ]
}
```
2. get label values of the specified label name, but label name != `__name__`. the result is correct
```
$ curl -s  "127.0.0.1:8481/select/0/prometheus/api/v1/label/t1/values" -d 'match[]={t1="v1"}' | jq .
{
  "status": "success",
  "data": [
    "v1"
  ]
}
```

3. get label values of the specified label name, but label name == `__name__`. the result is `incorrect`
```
$ curl -s  "127.0.0.1:8481/select/0/prometheus/api/v1/label/__name__/values" -d 'match[]={t1="v1"}' | jq .
{
  "status": "success",
  "data": []
}
```

**Expected behavior**
get the following result
```
{
  "status": "success",
  "data": [
    "m_v"
  ]
}
```

**Version**
```
$ ./bin/vmstorage --version
vmstorage-20200227-062002-heads-cluster-0-g52fe4e6

$ ./bin/vminsert --version
vminsert-20200227-061952-heads-cluster-0-g52fe4e6

$ ./bin/vmselect --version
vmselect-20200227-061958-heads-cluster-0-g52fe4e6

```

**Comment by valyala** [comment]
Confirmed the bug. It is reproduced only if non-empty `match[]` filter is passed to the query with `/api/v1/label/__name__/values`. I.e. it isn't reproduced with the following query:

```
$ curl -s  "127.0.0.1:8481/select/0/prometheus/api/v1/label/__name__/values" | jq .
{
  "status": "success",
  "data": [
    "m_v"
  ]
}
```

**Comment by valyala** [comment]
The issue should be fixed in the commit 8fa1cd24d8cc0d3cc6125e66377ff4ddf097af48 .

@n4mine , could you build `vmselect` from this commit and verify whether the issue is fixed? See [instructions on how to build cluster components from sources](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/cluster/README.md#building-from-sources).

**Comment by n4mine** [comment]
yeah, it's fixed.

---

But i found it's slowly if the labelName == `__name__`

1. labelName != `__name__`
``` 
$ for i in $(seq 10); do  /bin/time -f "%e" curl -so /dev/null 127.0.0.1:8481/select/0/prometheus/api/v1/label/port/values -d 'match[]={__ns__="foo"}' ; done
0.37
0.24
0.24
0.23
0.21
0.23
0.22
0.22
0.24
0.24
```

2. labelName == `__name__`
```
$ for i in $(seq 10); do  /bin/time -f "%e" curl -so /dev/null 127.0.0.1:8481/select/0/prometheus/api/v1/label/__name__/values -d 'match[]={__ns__="foo"}' ; done
2.20
1.07
1.01
1.03
0.26
0.91
1.06
0.29
1.04
1.06
```

3. without labelName filter
```
$ for i in $(seq 10); do  /bin/time -f "%e" curl -so /dev/null 127.0.0.1:8481/select/0/prometheus/api/v1/label/__name__/values ; done
0.01
0.01
0.01
0.01
0.01
0.01
0.01
0.01
0.01
0.01
```

And additional data:
1. series count
```
$ curl -s 127.0.0.1:8481/select/0/prometheus/api/v1/series/count | jq .
{
  "status": "success",
  "data": [
    29046846
  ]
}
```

2. labels count
```
$ curl -s 127.0.0.1:8481/select/0/prometheus/api/v1/labels/count | jq .
{
  "status": "success",
  "data": {
    "pod": 90784,
    "container": 59229,
    "ip": 50588,
    "reason": 38574,
    "host": 19475,
    "mount": 13288,
    "node": 13110,
    "nodeIP": 4753,
    "__name__": 2349,
    "name": 2279,
    "ns": 1463,
    "want": 739,
    "__ns__": 448,
    "key": 362,
......
```

**Comment by valyala** [comment]
Query time for `/api/v1/label/__name__/values` with non-empty `match[]` arg should be reduced in the commit d23df53ba28bd80bd17508586d01847c6d34dbda . @n4mine , could you verify this?

As for different query times between queries to `/api/v1/label/<label_name>/values` with and without `match[]` arg, this is because they take different code paths. When `match[]` arg is missing, the query is usually executed much faster, since it fetches label values from a small list. When `match[]` arg is set, VictoriaMetrics falls back to slower path, which requires fetching all the labels for all the time series matching the filter from `match[]` and then extracting the requested `<label_name>` from the results.

**Comment by n4mine** [comment]
Looks good. tks

```
$ for i in $(seq 10); do  /bin/time -f "%e" curl -so /dev/null 127.0.0.1:8481/select/0/prometheus/api/v1/label/__name__/values -d 'match[]={__ns__="foo"}' ; done
1.43
0.43
0.31
0.30
0.32
0.28
0.30
0.26
0.29
0.31
```

**Comment by valyala** [comment]
@n4mine , FYI, the bugfix has been included in [v1.32.4](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).
